pyhf/pdf.py

```diff
@@ -126,7 +128,44 @@ def __init__(self, spec, **config_kwargs):
         # build up our representation of the specification
         self.config = _ModelConfig.from_spec(self.spec,**config_kwargs)
 
+    def _mtype_results(self,mtype,pars):
+        mtype_results = {}
+        for channel in self.spec['channels']:
+            for sample in channel['samples']:
+                for mname in sample['modifiers_by_type'].get(mtype,[]):
+                    modifier = self.config.modifier(mname)
+                    modpars  = pars[self.config.par_slice(mname)]
+                    mtype_results.setdefault(channel['name'],
+                            {}).setdefault(sample['name'],
+                            []).append(
+                                modifier.apply(channel, sample, modpars)
+                            )
+        return mtype_results
+
     def expected_sample(self, channel, sample, pars):
+        tensorlib, _ = get_backend()
+        #public API only, not efficient
+        all_modifications = self._all_modifications(pars)
+        return self._expected_sample(
+            tensorlib.astensor(sample['data']), #nominal
+            *all_modifications[channel['name']][sample['name']] #mods
+        )
+
```

Comment:
the comment reflects the fact that we do not expect this to be fast. we compute *all* modification for all samples in this method even though we are only interested in the modifications in a single sample. Alternatively the `_all_modifications` could take a list of channel/samples for which it should compute mods